doc/source/exclude-patterns.md

```diff
@@ -0,0 +1,46 @@
+## Introduction
+
+The optional argument `exclude_patterns` in the `in_toto_run` API, also used by
+`--exclude` in the `in-toto-run` command line tool, applies
+[pathspec](http://python-path-specification.readthedocs.io) to compile
+[gitignore-style patterns](https://git-scm.com/docs/gitignore). Artifacts
+(materials and products) matched by an exclude pattern are not recorded when
+generating link metadata.
+
+## Pattern Formats
+
+- Single asterisks match everything except a slash.
+- Question marks match any one character except a slash.
+- Ranges of characters can be used in patterns using the `[]` notation (i.e.,
+  `[a-z]`).
+- A forward slash indicates a directory separator.
+- A single `/` does not match any file.
+- Separators at the beginning or middle (or both) of a pattern are relative to
+  the current directory. Separators at the end of the pattern only match
+  directories. (e.g., `doc/frotz/` matches the `doc/frotz/` directory but not
+  `a/doc/frotz/`; however `frotz/` matches `frotz/` and `a/frotz/`).
```

Comment:
> Separators at the beginning or middle (or both) of a pattern are relative to  the current directory.

Can we add an example with a separator at the beginning, and middle (and both) of a pattern?

>  Separators at the end of the pattern only match directories. (e.g., `doc/frotz/` matches the `doc/frotz/` directory but not `a/doc/frotz/`, however `frotz/` matches `frotz/` and `a/frotz/`.

I still don't understand the choice of the examples here. The rule doesn't really explain why  `doc/frotz/` doesn't match `a/doc/frotz/`. Also, "`doc/frotz/` matches `doc/frotz/`" seems redundant with "`frotz/` matches `frotz/`".

Reply:
> The rule doesn't really explain why doc/frotz/ doesn't match a/doc/frotz/.

That's a fair point, and it seems to be by design. Perhaps worth looking up the design decision by git / pathspec.

> Also, "doc/frotz/ matches doc/frotz/" seems redundant with "frotz/ matches frotz/".

No, actually, I think it's worth showing this because of the following:

```
>>> exclude_patterns = ["doc/frotz/"]
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns)
>>> _apply_exclude_patterns(["a/doc/frotz/", "doc/frotz/"], exclude_filter)
['a/doc/frotz/']
```

```
>>> exclude_patterns = ["frotz/"]
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns)
>>> _apply_exclude_patterns(["a/doc/frotz/", "doc/frotz/"], exclude_filter)
[]
```

Reply:
> Can we add an example with a separator at the beginning, and middle (and both) of a pattern?

I think `doc/frotz` highlights that actually, and that's why it's not redundant with `frotz/`. I'll add something for the beginning.

```
>>> exclude_patterns = ["/frotz/"]
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns)
>>> _apply_exclude_patterns(["a/doc/frotz/", "doc/frotz/"], exclude_filter)
['a/doc/frotz/', 'doc/frotz/']
>>> _apply_exclude_patterns(["a/doc/frotz/", "frotz/"], exclude_filter)
['a/doc/frotz/']
```

Reply:
Thanks for the clarifications! Just approved the PR. I think you need to approve it too. :)